src/frequenz/dispatch/actor.py

```diff
@@ -0,0 +1,199 @@
+            for task in self._scheduled.values():
+                task.cancel()
+            raise
+
+    async def _fetch(self) -> None:
+        """Fetch all relevant dispatches."""
+        old_dispatches = self._dispatches
+        self._dispatches = {}
+
+        try:
+            logging.info("Fetching dispatches for microgrid %s", self._microgrid_id)
+            async for dispatch in self._client.list(microgrid_id=self._microgrid_id):
+                self._dispatches[dispatch.id] = dispatch
+
+                old_dispatch = old_dispatches.pop(dispatch.id, None)
+                if not old_dispatch:
+                    self._update_dispatch_schedule(dispatch, None)
+                    logging.info("New dispatch: %s", dispatch)
+                    await self._updated_dispatch_sender.send(dispatch)
+                elif dispatch.update_time != old_dispatch.update_time:
+                    self._update_dispatch_schedule(dispatch, old_dispatch)
+                    logging.info("Updated dispatch: %s", dispatch)
+                    await self._updated_dispatch_sender.send(dispatch)
+
+        except grpc.aio.AioRpcError as error:
+            logging.error("Error fetching dispatches: %s", error)
+            self._dispatches = old_dispatches
+            return
+
+        for dispatch in old_dispatches.values():
+            logging.info("Deleted dispatch: %s", dispatch)
+            if task := self._scheduled.pop(dispatch.id, None):
+                task.cancel()
+
+    def _update_dispatch_schedule(
+        self, dispatch: Dispatch, old_dispatch: Dispatch | None
+    ) -> None:
+        """Update the schedule for a dispatch.
+
+        Schedules, reschedules or cancels the dispatch based on the start_time
+        and active status.
+
+        Args:
+            dispatch: The dispatch to update the schedule for.
+            old_dispatch: The old dispatch, if available.
+        """
+        if (
+            old_dispatch
+            and old_dispatch.active
+            and old_dispatch.start_time != dispatch.start_time
+        ):
+            if task := self._scheduled.pop(dispatch.id, None):
+                task.cancel()
+
+        if dispatch.active and dispatch.id not in self._scheduled:
+            self._scheduled[dispatch.id] = asyncio.create_task(
+                self._schedule_task(dispatch)
+            )
+
+    async def _schedule_task(self, dispatch: Dispatch) -> None:
+        """Wait for a dispatch to become ready.
+
+        Waits for the dispatches next run and then notifies that it is ready.
+
+        Args:
+            dispatch: The dispatch to schedule.
+        """
+        while True:
+            next_time = self.calculate_next_run(
+                dispatch, datetime.now().replace(tzinfo=timezone.utc)
+            )
+            if next_time is None:
+                logging.info("Dispatch finished: %s", dispatch)
+                self._scheduled.pop(dispatch.id)
+                return
```

Comment:
maybe more readable with:

```python
while next_time := self.calculate_next_run(dispatch, datetime.now(tz=timezone.utc)):
    ...

# and after the loop
logging.info("Dispatch finished: %s", dispatch)
self._scheduled.pop(dispatch.id)
```

Reply:
then I can't reuse `now` easily with your suggestion, but I modified it a bit..

```
        def next_run_info() -> tuple[datetime, datetime] | None:
            now = datetime.now(tz=timezone.utc)
            next_run = self.calculate_next_run(dispatch, now)

            if next_run is None:
                return None

            return now, next_run

        while pair := next_run_info():
            now, next_time = pair

            if next_time - now > _MAX_AHEAD_SCHEDULE:
                await asyncio.sleep(_MAX_AHEAD_SCHEDULE.total_seconds())
                continue

            _logger.info("Dispatch %s scheduled for %s", dispatch.id, next_time)
            _logger.info("Sleeping for %s", next_time - now)
            await asyncio.sleep((next_time - now).total_seconds())

            _logger.info("Dispatch ready: %s", dispatch)
            await self._ready_dispatch_sender.send(dispatch)

        _logger.info("Dispatch finished: %s", dispatch)
        self._scheduled.pop(dispatch.id)

```